File: navigation/src/main/java/com/example/navigation/stepsEngine/flow/rules/comparable/ComparableRule.java
package com.example.navigation.stepsEngine.flow.rules.comparable;


import com.example.navigation.stepsEngine.field.FlowStateField;
import com.example.navigation.stepsEngine.flow.rules.base.Rule;

import java.lang.reflect.Type;

public abstract class ComparableRule extends Rule {

//    private FieldManager fieldManager;
//
//    public FieldManager getFieldManager() {
//        return fieldManager;
//    }
//
//    public void setFieldManager(FieldManager fieldManager) {
//        this.fieldManager = fieldManager;
//    }


    FlowStateField field;
    String value;
    String valueType;

    public boolean isNot() {
        return not;
    }

    boolean not;


    public String getFieldName() {
        return field.getFieldName();
    }

    public String getValue() {
        return value;
    }

    public String getValueType() {
        return valueType;
    }

    public FlowStateField getField(){
        return field;
    }


    public Type getRuleClass(){
        if (getValueType().equals("bool"))
            return Boolean.class;
        else if (getValueType().equals("string"))
             return String.class;
        else if (getValueType().equals("decimal"))
            return Double.class;
        else if (getValueType().equals("int"))
            return Integer.class;
        else
            throw new IllegalStateException(String.format("Error in value_type. Check your json!!!"));
    }


}

File: navigation/src/main/java/com/example/navigation/stepsEngine/flow/rules/comparable/ComparableRule.kt
package com.example.navigation.stepsEngine.flow.rules.comparable

import com.example.navigation.action.FlowManager
import com.example.navigation.stepsEngine.field.Field
import com.example.navigation.stepsEngine.flow.rules.base.Rule
import com.example.navigation.stepsEngine.payment.FlowState

abstract class ComparableRule : Rule() {

    protected open lateinit var field: String
    open lateinit var value: String
    var valueType: String? = null
    var isNot = false
    private var fieldValue: Field? = null


    private fun getField(flowState: FlowState): Field? {
        if (fieldValue == null)
            fieldValue = flowState.getField(field)
        return fieldValue;
    }

    fun getValue(flowState: FlowState): Any? {
        val value = getField(flowState)!!.get(flowState)
        if (this !is NullRule)
            checkNotNull(value) { String.format("the %s field must not be null", getField(flowState)!!.id)}
        return value
    }

}